### PR TITLE
iknv_obj ok

### PR DESCRIPTION
**Pull request by kyleburnette**
Before opening this PR, ensure the following:
- `./format.sh` was run to apply standard formatting.
- `make` successfully builds a matching ROM.
- No new compiler warnings were introduced during the build process.
    - Can be verified locally by running `tools/warnings_count/check_new_warnings.sh`
- New variables & functions should follow standard naming conventions.
- Comments and variables have correct spelling.
---

OK, but I'm honestly not 100% sure what this actor is. Cloudmodding says "Ikana Canyon Objects", which I guess is fair enough considering this actor seems to handle ~3 different types of objects. On our MM actor sheet, someone filled it in as "Music Box House - Waterwheel" which could be one of the objects 

Anyway, if anyone has documentation ideas or anything I'm all ears. I'll bust out kz and check it out soon.

### CODE REVIEW

**Comment by kyleburnette** [comment]
Did some digging and found that type 0 is the waterwheel, type 1 is a door that seemingly used to cover the entrance to stone tower from ikana, and type 2 is sakon's hideout door.

File: src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c [hunk 1]
```diff
@@ -21,32 +31,190 @@ const ActorInit Bg_Iknv_Obj_InitVars = {
...
+        case IKNV_OBJ_SAKON_DOOR: {
+            this->displayListPtr = D_060129C8;
+            this->actionFunc = BgIknvObj_UpdateSakonDoor;
+            BcCheck3_BgActorInit(&this->dyna, 0);
+            BgCheck_RelocateMeshHeader(&D_06012CA4, &colHeader);
+            this->dyna.bgId = BgCheck_AddActorMesh(globalCtx, &globalCtx->colCtx.dyna, &this->dyna, colHeader);
+            Collider_InitAndSetCylinder(globalCtx, &this->collider, &this->dyna.actor, &sCylinderInit);
+            Collider_UpdateCylinder(&this->dyna.actor, &this->collider);
+            this->dyna.actor.colChkInfo.mass = 0xFF;
+            gSaveContext.weekEventReg[0x33] &= 0xEF;
```
Comment (engineer124) [inline src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c:88]:
```suggestion
            this->dyna.actor.colChkInfo.mass = MASS_IMMOVABLE;
            gSaveContext.weekEventReg[51] &= ~0x10;
```

File: src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c [hunk 2]
```diff
@@ -21,32 +31,190 @@ const ActorInit Bg_Iknv_Obj_InitVars = {
...
+    Actor_MarkForDeath(&this->dyna.actor);
+}
+
+void BgIknvObj_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+    BgIknvObj* this = THIS;
+
+    if (IKNV_OBJ_TYPE(this) != IKNV_OBJ_RAISED_DOOR) {
+        if (IKNV_OBJ_TYPE(this) == IKNV_OBJ_SAKON_DOOR) {
+            Collider_DestroyCylinder(globalCtx, &this->collider);
+            gSaveContext.weekEventReg[0x33] &= 0xEF;
```
Comment (engineer124) [inline src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c:103]:
```suggestion
            gSaveContext.weekEventReg[51] &= ~0x10;
```

File: src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c [hunk 3]
```diff
@@ -21,32 +31,190 @@ const ActorInit Bg_Iknv_Obj_InitVars = {
...
+        return 1;
+    }
+
+    if (ActorCutscene_GetCanPlayNext(this->dyna.actor.cutscene)) {
+        ActorCutscene_StartAndSetUnkLinkFields(this->dyna.actor.cutscene, &this->dyna.actor);
+        return 1;
+    }
+
+    ActorCutscene_SetIntentToPlay(this->dyna.actor.cutscene);
+    return 0;
```
Comment (engineer124) [inline src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c:126]:
```suggestion
        return true;
    }

    if (ActorCutscene_GetCurrentIndex() == this->dyna.actor.cutscene) {
        return true;
    }

    if (ActorCutscene_GetCanPlayNext(this->dyna.actor.cutscene)) {
        ActorCutscene_StartAndSetUnkLinkFields(this->dyna.actor.cutscene, &this->dyna.actor);
        return true;
    }

    ActorCutscene_SetIntentToPlay(this->dyna.actor.cutscene);
    return false;
```

File: src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c [hunk 4]
```diff
@@ -21,32 +31,190 @@ const ActorInit Bg_Iknv_Obj_InitVars = {
...
+}
+
+void BgIknvObj_UpdateWaterwheel(BgIknvObj* this, GlobalContext* globalCtx) {
+    if (gSaveContext.weekEventReg[0xE] & 4) {
```
Comment (engineer124) [inline src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c:130]:
```suggestion
    if (gSaveContext.weekEventReg[14] & 4) {
```

File: src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c [hunk 5]
```diff
@@ -21,32 +31,190 @@ const ActorInit Bg_Iknv_Obj_InitVars = {
...
+        this->dyna.actor.shape.rot.z += -0x64;
```
Comment (engineer124) [inline src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c:131]:
```suggestion
        this->dyna.actor.shape.rot.z -= 0x64;
```
I know negatives can be finicky sometimes, does this work?

File: src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c [hunk 6]
```diff
@@ -21,32 +31,190 @@ const ActorInit Bg_Iknv_Obj_InitVars = {
...
+        func_800B9098(&this->dyna.actor);
+        func_800B9010(&this->dyna.actor, 0x21AA);
```
Comment (engineer124) [inline src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c:133]:
```suggestion
        func_800B9010(&this->dyna.actor, NA_SE_EV_WOOD_WATER_WHEEL - SFX_FLAG);
```

File: src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c [hunk 7]
```diff
@@ -21,32 +31,190 @@ const ActorInit Bg_Iknv_Obj_InitVars = {
...
+    }
+
+    if (globalCtx->csCtx.state != 0 && gSaveContext.sceneSetupIndex == 1 && globalCtx->csCtx.unk12 == 4 &&
+        globalCtx->csCtx.frames == 0x5D7) {
+        func_8019F128(0x29A9);
```
Comment (engineer124) [inline src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c:138]:
```suggestion
        func_8019F128(NA_SE_EV_DOOR_UNLOCK);
```

File: src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c [hunk 8]
```diff
@@ -21,32 +31,190 @@ const ActorInit Bg_Iknv_Obj_InitVars = {
...
+    }
+}
+
+s32 func_80BD7E0C(BgIknvObj* this, s16 targetRotation, GlobalContext* globalCtx) {
+    this->dyna.actor.shape.yOffset = 0.0f;
+    CollisionCheck_SetOC(globalCtx, &globalCtx->colCheckCtx, &this->collider.base);
+    if (targetRotation != this->dyna.actor.shape.rot.y) {
+        Math_SmoothStepToS(&this->dyna.actor.shape.rot.y, targetRotation, 2, 100, 100);
+        this->dyna.actor.world.rot.y = this->dyna.actor.shape.rot.y;
+        if ((globalCtx->gameplayFrames & 1) != 0) {
```
Comment (engineer124) [inline src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c:148]:
```suggestion
        if ((globalCtx->gameplayFrames % 2) != 0) {
```
Would this work?

File: src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c [hunk 9]
```diff
@@ -21,32 +31,190 @@ const ActorInit Bg_Iknv_Obj_InitVars = {
...
+            this->dyna.actor.shape.yOffset = 5.0f;
+        }
+        func_800B9010(&this->dyna.actor, NA_SE_EV_STONEDOOR_OPEN_S - SFX_FLAG);
+        return 0;
+    }
+    Audio_PlayActorSound2(&this->dyna.actor, NA_SE_EV_STONEDOOR_STOP);
+    return 1;
```
Comment (engineer124) [inline src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c:155]:
```suggestion
        return false;
    }
    Audio_PlayActorSound2(&this->dyna.actor, NA_SE_EV_STONEDOOR_STOP);
    return true;
```

File: src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c [hunk 10]
```diff
@@ -21,32 +31,190 @@ const ActorInit Bg_Iknv_Obj_InitVars = {
...
+}
+
+void func_80BD7ED8(BgIknvObj* this, GlobalContext* globalCtx) {
+    if (func_80BD7E0C(this, this->dyna.actor.home.rot.y, globalCtx)) {
+        this->actionFunc = BgIknvObj_UpdateSakonDoor;
+        gSaveContext.weekEventReg[0x33] &= 0xEF;
```
Comment (engineer124) [inline src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c:161]:
```suggestion
        gSaveContext.weekEventReg[51] &= ~0x10;
```

File: src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c [hunk 11]
```diff
@@ -21,32 +31,190 @@ const ActorInit Bg_Iknv_Obj_InitVars = {
...
+    }
+    CollisionCheck_SetOC(globalCtx, &globalCtx->colCheckCtx, &this->collider.base);
+}
+
+void func_80BD7F4C(BgIknvObj* this, GlobalContext* globalCtx) {
+    if (gSaveContext.time >= 0xD001) {
```
Comment (engineer124) [inline src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c:167]:
```suggestion
    if (gSaveContext.time > 0xD000) {
```

File: src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c [hunk 12]
```diff
@@ -21,32 +31,190 @@ const ActorInit Bg_Iknv_Obj_InitVars = {
...
+        this->actionFunc = func_80BD7ED8;
+    }
+    if (this->dyna.actor.home.rot.x == 1 && (gSaveContext.weekEventReg[0x3A] & 0x80) == 0) {
```
Comment (engineer124) [inline src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c:170]:
```suggestion
    if (this->dyna.actor.home.rot.x == 1 && !(gSaveContext.weekEventReg[58] & 0x80)) {
```

File: src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c [hunk 13]
```diff
@@ -21,32 +31,190 @@ const ActorInit Bg_Iknv_Obj_InitVars = {
...
+        ActorCutscene_Stop(this->dyna.actor.cutscene);
+        this->dyna.actor.home.rot.x = 0;
+    }
+    CollisionCheck_SetOC(globalCtx, &globalCtx->colCheckCtx, &this->collider.base);
+}
+
+void func_80BD7FDC(BgIknvObj* this, GlobalContext* globalCtx) {
+    if (func_80BD7E0C(this, this->dyna.actor.home.rot.y + 0x4000, globalCtx)) {
+        this->actionFunc = func_80BD7F4C;
+        gSaveContext.weekEventReg[0x33] |= 0x10;
```
Comment (engineer124) [inline src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c:180]:
```suggestion
        gSaveContext.weekEventReg[51] |= 0x10;
```

File: src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c [hunk 14]
```diff
@@ -21,32 +31,190 @@ const ActorInit Bg_Iknv_Obj_InitVars = {
...
+    if (func_80BD7CEC(this)) {
+        this->actionFunc = func_80BD7FDC;
+    }
+    CollisionCheck_SetOC(globalCtx, &globalCtx->colCheckCtx, &this->collider.base);
+}
+
+void BgIknvObj_UpdateSakonDoor(BgIknvObj* this, GlobalContext* globalCtx) {
+    if ((gSaveContext.weekEventReg[0x3A] & 0x80) != 0) {
+        this->actionFunc = func_80BD8040;
+        gSaveContext.weekEventReg[0x59] |= 0x80;
```
Comment (engineer124) [inline src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c:195]:
```suggestion
    if (gSaveContext.weekEventReg[58] & 0x80) {
        this->actionFunc = func_80BD8040;
        gSaveContext.weekEventReg[89] |= 0x80;
```

File: src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c [hunk 15]
```diff
@@ -21,32 +31,190 @@ const ActorInit Bg_Iknv_Obj_InitVars = {
...
+            Collider_InitAndSetCylinder(globalCtx, &this->collider, &this->dyna.actor, &sCylinderInit);
+            Collider_UpdateCylinder(&this->dyna.actor, &this->collider);
+            this->dyna.actor.colChkInfo.mass = 0xFF;
+            gSaveContext.weekEventReg[0x33] &= 0xEF;
+            Actor_SetHeight(&this->dyna.actor, IREG(88));
+            return;
+        }
+    }
+
+    Actor_MarkForDeath(&this->dyna.actor);
```
Comment (engineer124) [inline src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c:94]:
```suggestion
            break;
        }
        default:
            Actor_MarkForDeath(&this->dyna.actor);
    }  
```
Would it still match if you change all the `return;` into `break;` and put `Actor_MarkForDeath` in `default:`?

File: src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c [hunk 16]
```diff
@@ -21,32 +31,190 @@ const ActorInit Bg_Iknv_Obj_InitVars = {
...
+    Actor_SetScale(&this->dyna.actor, 0.1f);
+    this->actionFunc = BgIknvObj_DoNothing;
+    switch (IKNV_OBJ_TYPE(this)) {
+        case IKNV_OBJ_WATERWHEEL: {
+            this->displayListPtr = D_06013058;
+            this->actionFunc = BgIknvObj_UpdateWaterwheel;
+            this->dyna.actor.flags |= 0x100000;
+            this->dyna.actor.flags |= 0x10;
+            return;
+        }
```
Comment (engineer124) [inline src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c:69]:
```suggestion
        case IKNV_OBJ_WATERWHEEL:
            this->displayListPtr = D_06013058;
            this->actionFunc = BgIknvObj_UpdateWaterwheel;
            this->dyna.actor.flags |= 0x100000;
            this->dyna.actor.flags |= 0x10;
            return;
```
You don't need curly braces between each case. Only one for the entire switch

File: src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c [hunk 17]
```diff
@@ -21,32 +31,190 @@ const ActorInit Bg_Iknv_Obj_InitVars = {
...
+
+void BgIknvObj_UpdateWaterwheel(BgIknvObj* this, GlobalContext* globalCtx) {
+    if (gSaveContext.weekEventReg[0xE] & 4) {
+        this->dyna.actor.shape.rot.z += -0x64;
+        func_800B9098(&this->dyna.actor);
+        func_800B9010(&this->dyna.actor, 0x21AA);
+    }
+
+    if (globalCtx->csCtx.state != 0 && gSaveContext.sceneSetupIndex == 1 && globalCtx->csCtx.unk12 == 4 &&
+        globalCtx->csCtx.frames == 0x5D7) {
```
Comment (engineer124) [inline src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c:137]:
```suggestion
    if ((globalCtx->csCtx.state != 0) && (gSaveContext.sceneSetupIndex == 1) && (globalCtx->csCtx.unk12 == 4) &&
        (globalCtx->csCtx.frames == 0x5D7)) {
```

**Review by engineer124 — COMMENTED** [review]
Great work! Looks really good, mostly just little formatting comments

File: src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c [hunk 18]
```diff
@@ -21,32 +31,187 @@ const ActorInit Bg_Iknv_Obj_InitVars = {
...
+
+void func_80BD8040(BgIknvObj* this, GlobalContext* globalCtx) {
+    if (func_80BD7CEC(this)) {
+        this->actionFunc = func_80BD7FDC;
+    }
+    CollisionCheck_SetOC(globalCtx, &globalCtx->colCheckCtx, &this->collider.base);
+}
+
+void BgIknvObj_UpdateSakonDoor(BgIknvObj* this, GlobalContext* globalCtx) {
+    if ((gSaveContext.weekEventReg[58] & 0x80)) {
```
Comment (engineer124) [inline src/overlays/actors/ovl_Bg_Iknv_Obj/z_bg_iknv_obj.c:190]:
```suggestion
    if (gSaveContext.weekEventReg[58] & 0x80) {
```

**Review by engineer124 — APPROVED** [review]
Looks good to me! Caught the smallest extra brackets, but everything looks fine

**Review by Kenix3 — APPROVED** [review]

**Pull request merged**